### PR TITLE
fix(builtin/type): a plain array never matches a map type

### PR DESCRIPTION
**Pull request by mck**
Part of #142 (item 7).

## What

`matchesMapType` and `typeAssert`'s inline map-type branch accepted any
non-null object as a candidate map, including a plain JS `Array`.

## Why

`Object.entries()` of an array yields string-indexed pairs
(`[['0', v0], ['1', v1], ...]`), which pass the key/elem sampling used to
structurally match a map type. So a Go type switch that lists
`case map[string]any:` before `case []any:` captures slices in the map
branch, and ranging over the result calls `Array.prototype.entries()`,
producing numeric-keyed output instead of correct slice semantics:

```go
var v any = []any{"a", "b"}
switch x := v.(type) {
case map[string]any:
    // wrongly taken; x behaves like {0: "a", 1: "b"}
case []any:
    // never reached
}
```

A Go slice is never a map, so both `matchesMapType` and `typeAssert`'s map
branch now reject `Array`/`Uint8Array` values up front before doing any
key/elem sampling — no further type information is needed to know a slice
isn't a map.

## Fix

Two mechanical guards, added right after the existing null/object checks:

- `matchesMapType`: `if (Array.isArray(value) || value instanceof Uint8Array) return false`
- `typeAssert`'s inline map branch: same condition added to the guard so it
  falls through to the normal `matchesType` path (which now correctly
  rejects it via the fixed `matchesMapType`) instead of the special-cased
  map-assertion logic.

## Test

Added `gs/builtin/type.test.ts` covering `is()`, `typeAssert()`, and
`typeSwitch()` against `map[string]any` vs `[]any`, plus the empty-value
edge case: an empty `map[string]any` check previously matched an empty
array too, via the "empty map matches any map type" shortcut in both
functions.

All cases fail on `master` and pass with this fix.

## Verify

```
bun run vitest run gs/builtin/type.test.ts
```
7 passed.

Full suite:
- `bun run typecheck` — clean
- `bun run test:js` (typecheck + vitest) — 1278 passed, 4 skipped, 0 failed
- `go test ./...` — all packages green
- `bun run lint:js` — clean

Found while transpiling and running a real ~100k-LOC Go interpreter/parser
package through goscript.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: gs/builtin/type.test.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+// []any
+const sliceAny: SliceTypeInfo = {
+  kind: TypeKind.Slice,
+  elemType: { kind: TypeKind.Interface, methods: [] },
+}
+
+describe('a Go slice is never a map (structural type matching)', () => {
+  it('is(): a plain array does not match map[string]any', () => {
+    expect(is([1, 2, 3], mapStringAny)).toBe(false)
+  })
```
[the remark this answers is not in the document]
Comment (mck) [reply gs/builtin/type.test.ts:28]: Good catch — added a Uint8Array case alongside the plain-array is()/typeAssert() rejection tests in 15b63454.

File: gs/builtin/type.test.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    )
+    expect(result).toBe('slice')
+  })
+
+  it('an empty array does not vacuously match an empty-map fast path', () => {
+    // matchesMapType/typeAssert's map branch special-cases zero entries as
+    // "matches any map type"; an empty slice must not hit that path.
+    expect(is([], mapStringAny)).toBe(false)
+    expect(is([], sliceAny)).toBe(true)
+  })
```
[the remark this answers is not in the document]
Comment (mck) [reply gs/builtin/type.test.ts]: Good catch — added an explicit empty-Uint8Array assertion to the empty-map-fast-path test in 15b63454.

**Review by paralin — APPROVED** [review]
Looks good. I verified the current head is still `15b63454`, including the extra `Uint8Array` coverage added after Copilot's comments.

Rejecting `Array` and `Uint8Array` before the map structural matcher is the right fix: a Go slice or byte slice is never a map, so the runtime can reject those shapes before key/element sampling.

Focused check:

```text
bun run vitest run gs/builtin/type.test.ts
# 7 passed
```
